Support all of HTML's character entities in WebVTT

cola119 opened the pull request:
#### 023c54054092dc68c5df3b230ed3137cbd753b16
<pre>
Support all of HTML&apos;s character entities in WebVTT
<a href="https://bugs.webkit.org/show_bug.cgi?id=176225">https://bugs.webkit.org/show_bug.cgi?id=176225</a>

Reviewed by Darin Adler.

WebVTT cue text tokenizer algorithm has been updated to support
all of HTML&apos;s character entities in <a href="https://github.com/w3c/webvtt/pull/253.">https://github.com/w3c/webvtt/pull/253.</a>
This patch updates the tokenizer to align with the latest spec.
Spec: <a href="https://w3c.github.io/webvtt/#webvtt-cue-text-tokenizer">https://w3c.github.io/webvtt/#webvtt-cue-text-tokenizer</a>

The old `EscapeState` state for handling escape characters has been
removed in favor of new two states. And the `WEBVTT_SWITCH_TO` operation
has been added for the purpose of enabling state transition without
the requirement to advance the input position.

* LayoutTests/imported/w3c/web-platform-tests/webvtt/parsing/cue-text-parsing/tests/entities-expected.txt:
* Source/WebCore/html/track/WebVTTTokenizer.cpp:
(WebCore::ProcessEntity):
(WebCore::WebVTTTokenizer::nextToken):

Canonical link: <a href="https://commits.webkit.org/270240@main">https://commits.webkit.org/270240@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/256d17ff8be37d8154089884bf8faf30fecde728

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/24962 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/3505 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26215 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27078 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/22914 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/25230 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/5198 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/944 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23196 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25206 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/2533 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21539 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/27658 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/2230 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/22475 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/28608 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/22764 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/22827 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26432 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2184 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/482 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/3465 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5973 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/2629 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/2526 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/html/track/WebVTTTokenizer.cpp:
```diff
@@ -209,7 +186,9 @@ bool WebVTTTokenizer::nextToken(WebVTTToken& token)
-    if (character == '>' || character == kEndOfFileMarker)
+    if (character == '&') {
+        WEBVTT_ADVANCE_TO(HTMLCharacterReferenceInAnnotationState);
+    } else if (character == '>' || character == kEndOfFileMarker)
```
In WebKit coding style this would not have braces. I suggest omitting them.

cola119 replied:
Done
https://github.com/WebKit/WebKit/compare/cec8ab3d2503e678a79df16e308353d9616529f1..256d17ff8be37d8154089884bf8faf30fecde728

cola119 committed:
Support all of HTML's character entities in WebVTT
https://bugs.webkit.org/show_bug.cgi?id=176225

Reviewed by Darin Adler.

WebVTT cue text tokenizer algorithm has been updated to support
all of HTML's character entities in https://github.com/w3c/webvtt/pull/253.
This patch updates the tokenizer to align with the latest spec.
Spec: https://w3c.github.io/webvtt/#webvtt-cue-text-tokenizer

The old `EscapeState` state for handling escape characters has been
removed in favor of new two states. And the `WEBVTT_SWITCH_TO` operation
has been added for the purpose of enabling state transition without
the requirement to advance the input position.

* LayoutTests/imported/w3c/web-platform-tests/webvtt/parsing/cue-text-parsing/tests/entities-expected.txt:
* Source/WebCore/html/track/WebVTTTokenizer.cpp:
(WebCore::ProcessEntity):
(WebCore::WebVTTTokenizer::nextToken):

Canonical link: https://commits.webkit.org/270240@main
